docs/ref/csp.txt

```diff
@@ -0,0 +1,296 @@
+
+    This decorator marks a view as being exempt from the Content Security Policy
+    protection ensured by the middleware. Example::
+
+        from django.http import HttpResponse
+        from django.views.decorators.csp import csp_exempt
+
+
+        @csp_exempt
+        def my_view(request):
+            return HttpResponse("This view will not have either CSP headers")
+
+    This decorator takes optional boolean arguments ``enforced`` and
+    ``report_only`` to customize which header is exempt. By default, both
+    headers are exempt. Example::
+
+        from django.http import HttpResponse
+        from django.views.decorators.csp import csp_exempt
+
+
+        @csp_exempt(report_only=False)
+        def my_view(request):
+            return HttpResponse("This view will only have an enforced CSP header")
+
+
+        @csp_exempt(enforced=False)
+        def another_view(request):
+            return HttpResponse("This view will only have a report-only CSP header")
+
+
+.. function:: csp_override(config, enforced=True, report_only=True)(view)
+
+    This decorator allows you to override the Content Security Policy for a
+    specific view.  It takes a dictionary as an argument, similar to the
+    ``SECURE_CSP`` and ``SECURE_CSP_REPORT_ONLY`` settings. This decorator also
+    takes optional boolean arguments ``enforced`` and ``report_only`` to
+    customize which header is overridden. By default, both headers are
+    overridden. Example::
+
+        from django.http import HttpResponse
+        from django.middleware import csp
+        from django.views.decorators.csp import csp_override
+
+
+        @csp_override(
+            {
+                "default-src": [csp.SELF],
+                "img-src": [csp.SELF, "data:"],
+            }
+        )
+        def my_view(request):
+            return HttpResponse("Both CSP headers will be overridden")
+
+
+        @csp_override(
+            {
+                "default-src": [csp.SELF],
+                "img-src": [csp.SELF, "data:"],
+            },
+            report_only=False,
+        )
+        def enforced_only_view(request):
+            return HttpResponse("Only the enforced CSP header will be overridden")
+
+
+        @csp_override(
+            {
+                "default-src": [csp.SELF],
+                "img-src": [csp.SELF, "data:"],
+            },
+            enforced=False,
+        )
+        def report_only_view(request):
+            return HttpResponse("Only the report-only CSP header will be overridden")
+
+These decorators provide fine-grained control over the Content Security Policy
+on a per-view basis, allowing you to customize the policy as needed for specific
+parts of your application.
+
+Limitations and Considerations
+==============================
+
+While Content Security Policy (CSP) is a powerful security feature, it's
+important to be aware of its limitations and considerations, especially in the
+context of Django's implementation:
```

Comment:
Great chapter! A bit long, but I get the intention and the ordering is good to have! Maybe a note about alternatively implementing CSP in the HTTP server layer?

Do you think this is too much _explanation_? I would think it's a tad heavy, so could consider moving it to the Security Topic. Other topics could be allowed to expand in the same fashion IMO.

Reply:
Which part is heavy? Just this section or the overall csp.txt file?

I built it with Sphinx and opened the HTML and it feels okay to me. I assume readers scan the headers mostly and dig in where needed. But open to suggestions on better structure or reducing text.

Reply:
I meant heavy on the explanation side. That's why I suggested that it could be a better fit over in the "security topics" section. I think the other topics there would be fine to expand in similar fashion to this chapter (considerations, limitations etc).

Reply:
I thought the ref/csp doc was where we'd include more of the details and fleshed out explanations, and the ref/security doc was meant to be a summarized version of all security features in Django? When someone searches "django csp" they will likely land here rather than the security page, or at least that's the top result for me when I search CSRF -- I get the ref/csrf doc.

I'm happy to move things around -- I need to better understand the motivation and goals behind it.

Reply:
I agree let's move this as a subheading within the CSP section for security